Enable GPU pipeline schedule with shared memory stores in stage 0

qedawkins opened the pull request:
Depends on #11086. Enables putting vector.transfer_write ops in stage 0 of the pipeline along with reading from global memory.

antiagainst commented on compiler/src/iree/compiler/Codegen/SPIRV/test/lowering_matmul_promotion.mlir:
```diff
@@ -51,34 +51,49 @@ hal.executable @matmul_f32_128x256x64 {
+//           CHECK:   memref.alloc() : memref<3x64x20xf32, 3>
+//           CHECK:   memref.alloc() : memref<3x16x68xf32, 3>
+//   CHECK-COUNT-8:   vector.transfer_write %{{.+}}, %{{.+}} {in_bounds = [true]} : vector<4xf32>, memref<128x256xf32>
+//   CHECK-COUNT-8:   vector.transfer_read %{{.+}}, %[[CST0]] {in_bounds = [true]} : memref<128x256xf32>, vector<4xf32>
```
Hmm, I expect the `vector.transfer_write` generated from `linalg.fill` should be forwarded to the `vector.transfer_read` here.. Do you know why that's not happening? Would be nice to figure out why. Not exactly for this pull request, but would be nice to at least put a TODO here.

antiagainst commented on compiler/src/iree/compiler/Codegen/SPIRV/test/lowering_matmul_promotion.mlir:
```diff
@@ -51,34 +51,49 @@ hal.executable @matmul_f32_128x256x64 {
+//           CHECK:   vector.transfer_write %{{.+}}, %{{.+}} {__pipelining_first_stage__, in_bounds = [true]} : vector<4xf32>, memref<3x64x20xf32, 3>
+//           CHECK:   vector.transfer_read %{{.+}}, %[[CST0]] {__pipelining_first_stage__, in_bounds = [true]} : memref<512x256xf32>, vector<4xf32>
+//           CHECK:   vector.transfer_write %{{.+}}, %{{.+}} {__pipelining_first_stage__, in_bounds = [true]} : vector<4xf32>, memref<3x16x68xf32, 3>
+//           CHECK:   gpu.barrier {__pipelining_first_stage__}
```
IIUC, we should be writing to different slices of the workgroup memory before and after this barrier. So is this barrier necessary here? @ThomasRaoux

qedawkins replied:
yeah they're writing to different slices, but I thought I removed this barrier before pipelining so I'm not yet sure why it ends up here.

antiagainst commented on compiler/src/iree/compiler/Codegen/SPIRV/KernelConfig.cpp:
```diff
@@ -575,14 +623,21 @@ LogicalResult setMatmulOpConfig(spirv::ResourceLimitsAttr limits,
+  auto storeStage = softwarePipelineStoreStage;
+
+  // No multi-buffering if fused with a non-matmul op.
+  if (isFusedWithUnbufferableOp(op)) {
```
Can we add a test for this case?

qedawkins replied:
Ah this comment is from before I figured out why it was failing for a specific case. I can add the above mlir as a test.

antiagainst replied:
+1

qedawkins replied:
This ended up being unnecessary after removing distribution loops before multi-buffering

antiagainst commented on compiler/src/iree/compiler/Codegen/SPIRV/test/lowering_matmul_promotion.mlir:
```diff
@@ -51,34 +51,49 @@ hal.executable @matmul_f32_128x256x64 {
 
-//       CHECK-DAG: #[[MAP:.+]] = affine_map<(d0) -> ((d0 floordiv 16) mod 2)>
-//       CHECK-DAG: #[[TRANSLATION:.+]] = #iree_codegen.translation_info<SPIRVMatmulPromoteVectorize pipeline_depth = 2>
+//       CHECK-DAG: #[[MAP:.+]] = affine_map<(d0) -> ((d0 floordiv 16) mod 3)>
```
Would it be easy to actually add a new test to check the case with the store_stage = 0 case? That would make sure we don't omit the previous pass in checks and only have the new case covered. If it's hard to think of a matmul shape fitting into the current heuristics, I'm thinking maybe using user config `#iree_codegen.compilation_info` to force it.

qedawkins replied:
Do you mean a test for the store_stage = 1 case? This is the store_stage = 0 case, but yeah I can add back in the previous case if that's what you mean.

antiagainst replied:
Yes sorry I meant store_stage = 1 case (which was the previous revision).

antiagainst commented on compiler/src/iree/compiler/Codegen/SPIRV/Passes.cpp:
```diff
@@ -335,11 +339,13 @@ void addSPIRVMatmulPromoteVectorizePassPipeline(OpPassManager &pm,
-  nestedModulePM.addNestedPass<func::FuncOp>(
-      createGPUPipeliningPass(pipelineDepth ? pipelineDepth : 1));
+  nestedModulePM.addNestedPass<func::FuncOp>(createGPUPipeliningPass(
+      /*epiloguePeeling=*/true, pipelineDepth, storeStage));
```
Might want to run canonicalization and cse pass to clean up stuff a bit after this? Or preferably, do it at the end of `addLoopMaterializationPasses` for all pipelines. (I suspect it will also make `CHECK` matching in tests easier.)

antiagainst commented on compiler/src/iree/compiler/Codegen/SPIRV/test/lowering_matmul_promotion.mlir:
```diff
@@ -51,34 +51,49 @@ hal.executable @matmul_f32_128x256x64 {
+//  CHECK-COUNT-16:   vector.transfer_read %{{.+}}, %[[CST0]] {in_bounds = [true]} : memref<3x16x68xf32, 3>, vector<4xf32>
 // CHECK-COUNT-128:   vector.fma %{{.+}}, %{{.+}}, %{{.+}} : vector<4xf32>
-//   CHECK-COUNT-8:   vector.transfer_read %{{.+}}, %[[CST0]] {in_bounds = [true]} : memref<8x4xf32, strided<[256, 1], offset: ?>>, vector<4xf32>
+//   CHECK-COUNT-8:   vector.transfer_write %{{.+}}, %{{.+}} {in_bounds = [true]} : vector<4xf32>, memref<128x256xf32>
```
Similar here. Shouldn't these writes be forwarded to the following reads? Would be nice to know why not. At least a TODO here for now.

antiagainst reviewed (changes requested):
Awesome, thanks for adding this! Overall LGTM; I just have a few questions and requests around tests. It would be great to get @ThomasRaoux's review on this too given he is the expert on this part.

qedawkins commented on compiler/src/iree/compiler/Codegen/SPIRV/KernelConfig.cpp:
```diff
@@ -48,6 +49,37 @@ bool isMatmulOrBatchMatmul(linalg::LinalgOp linalgOp) {
+}
+
+// Check if the given linalg op is fused with another op that may block
+// multi-buffering
```
I found that if the matmul is fused with an elementwise extf it will result in an alloc that multi-buffering can't find a loop that dominate the uses of it. Something like this
```
builtin.module {
  func.func @forward(%arg0: tensor<1x4096x512xf16>, %arg1: tensor<1x512x4096xf16>) -> tensor<1x4096x4096xf32> {
    %cst = arith.constant 0.000000e+00 : f16
    %5 = tensor.empty() : tensor<1x4096x4096xf32>
    %6 = tensor.empty() : tensor<1x4096x4096xf16>
    %7 = linalg.fill ins(%cst : f16) outs(%6 : tensor<1x4096x4096xf16>) -> tensor<1x4096x4096xf16>
    %8 = linalg.batch_matmul ins(%arg0, %arg1 : tensor<1x4096x512xf16>, tensor<1x512x4096xf16>) outs(%7 : tensor<1x4096x4096xf16>) -> tensor<1x4096x4096xf16>
    %9 = linalg.generic {indexing_maps = [affine_map<(d0, d1, d2) -> (d0, d1, d2)>, affine_map<(d0, d1, d2) -> (d0, d1, d2)>], iterator_types = ["parallel", "parallel", "parallel"]} ins(%8 : tensor<1x4096x4096xf16>) outs(%5 : tensor<1x4096x4096xf32>) {
    ^bb0(%in: f16, %out: f32):
      %10 = arith.extf %in : f16 to f32
      linalg.yield %10 : f32
    } -> tensor<1x4096x4096xf32>
    return %9 : tensor<1x4096x4096xf32>
  }
}

```
I can turn this into a test if that helps.

antiagainst replied:
I guess it's because `arith.extf` extends element bitwidth so it requires a larger output buffer output (`tensor<1x4096x4096xf32>`) than the matmul generates (`tensor<1x4096x4096xf16>`). Bufferization will create an intermediate buffer for it then. This seems to be a case we should fix bufferization for. Though then this is not specific to `arith.extf`; I suspect `arith.extsi` or similar ones will trigger this too.

Yes having such a test case would be helpful. Together, can we change this to check `OpTrait::hasElementwiseMappableTraits` and 1 operand and 1 result, but operand and result type have different bitwidth type? Some comments explaining this would also be great.

qedawkins replied:
> I guess it's because arith.extf extends element bitwidth so it requires a larger output buffer.

Yeah that is what I found, but perhaps just checking bitwidth makes more sense then, although I suspect there could still be cases missing. Namely, there are examples where I see a fusion with truncf that doesn't fail multi-buffering but I didn't take enough time to figure out why, although it would make sense that extf/si would come up with an additional buffer.

qedawkins replied:
I had to add something similar back in because the total shared memory usage wasn't being calculated properly. If there is a suggestion for a better way to handle this that would be best, otherwise I left it as a TODO that can be fixed with bufferization or by properly computing the shared memory usage.

MaheshRavishankar commented:
I dont have an idea about the GPU pipelining setup, so Ill let Thomas and Lei review that. Changes to LoweringConfig look good to me.

qedawkins commented:
> I dont have an idea about the GPU pipelining setup, so Ill let Thomas and Lei review that. Changes to LoweringConfig look good to me.

TY for the response, I was going to ask for confirmation on that after looking into the remaining pipelining comments.

qedawkins commented:
I added a check to remove extra barriers in the prologue when pipelining so a review on that part is needed as well.  Additionally I moved the RemoveSingleIterationLoop pass to happen before multi-buffering.

antiagainst reviewed (approved):
Nice, thanks a ton for pushing this through!!

ThomasRaoux commented on compiler/src/iree/compiler/Codegen/Dialect/LoweringConfig.td:
```diff
@@ -113,18 +113,22 @@ def IREECodegen_TranslationInfoAttr :
     `<` `` $passPipeline
-    (`pipeline_depth` `=` $softwarePipelineDepth^)? `>`
+    (`pipeline_depth` `=` $softwarePipelineDepth^)?
+    (`store_stage` `=` $softwarePipelineStoreStage^)? `>`
```
That makes sense, I think in the future we will have a bunch of different schedules we will probably want to make it an enum with schedule names.

ThomasRaoux reviewed (approved):
LGTM
